### ISSUE TITLE
reportUnnecessaryComparison ignores assert_never inside control flow blocks

### ISSUE THREAD

**Issue by vmphase** [DetachHead/basedpyright#1865]
### Description

When calling `assert_never()` directly on a variable whose type is narrowed to `Never`, basedpyright suppresses `reportUnnecessaryComparison`: 
- [playground example (bare call)](https://basedpyright.com/?pyrightVersion=1.28.5&typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCGAzkQKZ4D6KpAbuQFD0AmpwUwAFAB4BcmK%2BAD5QiMEAEooAWgB8-GD3pRlUCARgBjABZReSlQY3FS8juMUHLKkKRgBXECigBGfVaNkRYsxauWb9o5QAExulh4mFL5%2BKsRklNR0INziQA)

However, if the exact same `assert_never()` call is placed inside a nested block (like `if True:` or inside a `try-except` block), which is technically equal to the one above, `reportUnnecessaryComparison` is reported anyway:
- [playground example (if-true)](https://basedpyright.com/?pyrightVersion=1.28.5&typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCGAzkQKZ4D6KpAbuQFD0AmpwUwAFAB4BcmK%2BAD5QiMEAEooAWgB8-GD3pRlUCARgBjABZReSlQY3FS8juMUHLKkKRgBXECigBGfVaNkRYsxauWb9o5QAExulh4mFL5%2BKkhsACogdqTRMZbEZJTUdCDc4kA)

**Comment by DetachHead** [comment]
i'm curious whether there's a real life example where you'd want to put any statement other than an `assert_never` call inside an unreachable match case

**Comment by vmphase** [comment]
> i'm curious whether there's a real life example where you'd want to put any statement other than an `assert_never` call inside an unreachable match case

```python
case _:
    if DEBUG:
        print(f"Unexpected value: {x!r}")
        assert_never(x)
```
The variable is still narrowed to `Never`, so the same call should produce the same result.

**Comment by DetachHead** [comment]
hmm, i would still expect `reportUnreachable` to be reported there, but i guess `reportUnnecessaryComparison` shouldn't

**Comment by vmphase** [comment]
> hmm, i would still expect `reportUnreachable` to be reported there, but i guess `reportUnnecessaryComparison` shouldn't

Same do I, but the latter one is also reported.
